runtime/dancebox/tests/common/xcm/core_buyer.rs

```diff
@@ -0,0 +1,410 @@
+                },
+            ]
+        );
+    });
+}
+
+fn assert_relay_order_event_not_emitted() {
+    type RuntimeEvent = <Rococo as Chain>::RuntimeEvent;
+
+    let events = <Rococo as Chain>::events();
+    for event in events {
+        match event {
+            RuntimeEvent::OnDemandAssignmentProvider(
+                parachains_assigner_on_demand::Event::OnDemandOrderPlaced { .. },
+            ) => {
+                panic!("Event should not have been emitted: {:?}", event);
+            }
+            _ => (),
+        }
+    }
+    // OnDemandOrderPlaced
+}
+
+fn get_parathread_tank_relay_address() -> AccountId32 {
+    let parathread_tank_in_relay = Dancebox::execute_with(|| {
+        let parathread_tank_multilocation = BuyCoretime::absolute_multilocation(
+            BuyCoretime::interior_multilocation(PARATHREAD_ID.into()),
+        );
+        let parathread_tank_in_relay =
+            <Rococo as RelayChain>::SovereignAccountOf::convert_location(
+                &parathread_tank_multilocation,
+            )
+            .expect("probably this relay chain does not allow DescendOrigin");
+
+        parathread_tank_in_relay
+    });
+    parathread_tank_in_relay
+}
+
+fn get_on_demand_base_fee() -> u128 {
+    Rococo::execute_with(|| {
+        let config = <Rococo as RococoRelayPallet>::Configuration::config();
+
+        config.on_demand_base_fee
+    })
+}
+
+fn set_on_demand_base_fee(on_demand_base_fee: u128) {
+    Rococo::execute_with(|| {
+        let mut config = <Rococo as RococoRelayPallet>::Configuration::config();
+        config.on_demand_base_fee = on_demand_base_fee;
+        <Rococo as RococoRelayPallet>::Configuration::force_set_active_config(config);
+    });
+}
+
+#[test]
+fn xcm_core_buyer_zero_balance() {
+    let parathread_tank_in_relay = get_parathread_tank_relay_address();
+    let balance_before = 0;
+
+    // Invariant: if balance_before < BUY_EXECUTION_COST, then balance_after == balance_before
+    do_test(balance_before);
+
+    // Receive XCM message in Relay Chain
+    Rococo::execute_with(|| {
+        type RuntimeEvent = <Rococo as Chain>::RuntimeEvent;
+        let balance_after =
+            <Rococo as RococoRelayPallet>::System::account(parathread_tank_in_relay.clone())
+                .data
+                .free;
+        assert_expected_events!(
+            Rococo,
+            vec![
+                RuntimeEvent::MessageQueue(pallet_message_queue::Event::Processed { success: false, .. }) => {},
+            ]
+        );
+        assert_relay_order_event_not_emitted();
+        assert_eq!(balance_before, balance_after);
+    });
+}
+
+#[test]
+fn xcm_core_buyer_only_enough_balance_for_buy_execution() {
+    let parathread_tank_in_relay = get_parathread_tank_relay_address();
+    let balance_before = BUY_EXECUTION_COST;
+
+    // Invariant: if balance_before >= BUY_EXECUTION_COST then balance_after <= (balance_before - BUY_EXECUTION_COST)
```

Comment:
```suggestion
    // Invariant: if balance_before >= BUY_EXECUTION_COST then balance_after <= (balance_before + BUY_EXECUTION_REFUND - BUY_EXECUTION_COST)
```

In this test the final balance is 0 because `BUY_EXECUTION_REFUND < ROCOCO_ED`, so the account gets the refund but it is immediatelly killed.